Add additional information to summary

raph5 opened the pull request:
Hello,
I’d like to work a bit on the website to improve its readability and appearance.
This is an initial proposal that focuses solely on easy readability improvements, but I’d be happy to work on additional cosmetic and usability changes if you’d like.

<img width="742" alt="Screenshot 2025-05-12 at 12 53 20" src="https://github.com/user-attachments/assets/3d299583-d109-43c7-b959-812331a8a875" />

athas commented:
I certainly welcome any contribution aimed at making the website better!

The little logos are nice. I would prefer a slightly greater visual distinction - at a glance they look too similar, as they are both circles with a vertical bar inside. Maybe the timeout should be an hourglass instead, as it has a very different silhouette.

The numbers I am less sure of. The score is a fairly meaningless (and unitless) number for indexing a colour function, and I am concerned that making the quantity explicit will encourage inferences that are really not justified. E.g. a tool with a score of *2x* is not really "two times better" than a score of *x*.

raph5 commented:
I really think that the graph would benefit from displaying an objective number instead of a color. Though I agree that score is not very expressive. Maybe the average runtime over the various instances of the problem is a better label to display.

I think the graph, as it is, is adapted to be displayed in the github readme, but a more information rich graph is better for the website.

Can I change the color scale to a more expressive red-yellow-green or purple-orange-yellow one?

samestep reviewed (approved):
@raph5 this is awesome, thanks for helping out with the website!

> This is an initial proposal that focuses solely on easy readability improvements, but I’d be happy to work on additional cosmetic and usability changes if you’d like.

Yeah, there are actually several open issues about the website if you want to take a look: https://github.com/gradbench/gradbench/issues?q=is%3Aissue%20state%3Aopen%20label%3Awebsite

For example, I see that in this PR you take the same _relative_ score that we use to choose the cell color, and display that as text. This is fine with me; I'm guessing you tried using the absolute score instead and it didn't work as well. See issue #293 for a suggestion about letting people hide certain columns of the table to be able to compare slower tools more easily.

---

> The numbers I am less sure of. The score is a fairly meaningless (and unitless) number for indexing a colour function, and I am concerned that making the quantity explicit will encourage inferences that are really not justified. E.g. a tool with a score of *2x* is not really "two times better" than a score of *x*.

@athas would it be helpful if we were to add more help text on the website explaining exactly how the score is calculated, perhaps including a sentence warning against drawing these sorts of inferences you're worried about, and encouraging people to click to see the more detailed charts?

Perhaps this additional text could be in tooltips that people click or tap or hover to see, to avoid cluttering the whole page with too much text.

---

> Though I agree that score is not very expressive. Maybe the average runtime over the various instances of the problem is a better label to display.

@raph5 could you clarify what you mean by this? Isn't that exactly how the score is already calculated? Or do you mean to show that actual value rather than its reciprocal (which how the score is calculated for most evals)? See also this file for more details on how it's currently calculated:

https://github.com/gradbench/gradbench/blob/27bfb07f4e1727685d403914a323b604d14d8910/crates/gradbench/src/stats.rs#L278-L287

---

> Can I change the color scale to a more expressive red-yellow-green or purple-orange-yellow one?

We had red-yellow-green originally but this raised issue #294 so we changed it to this blue scale in #299. @athas what are your thoughts on changing it back? Personally I prefer the current blue scale.

athas commented:
> @athas would it be helpful if we were to add more help text on the website explaining exactly how the score is calculated, perhaps including a sentence warning against drawing these sorts of inferences you're worried about, and encouraging people to click to see the more detailed charts?

It's possible I'm just too cautious. I would be in favour of letting @raph5 put in the numbers. I am generally in favour of letting the people who do the work be the tiebreakers. I suppose time will tell if we will have a horde of confused users who don't understand the numbers. It could also encourage us to make sure the numbers are good (although I'm not sure how to make them better than what they currently are).

> @raph5 could you clarify what you mean by this? Isn't that exactly how the score is already calculated?

It is not for all evals (particle/saddle is the *best* runtime of the four variants), and as we eventually add more exotic evals, I imagine the meaning of the score will become very eval dependent (e.g. training quality within given time period).

> @athas what are your thoughts on changing it back? Personally I prefer the current blue scale.

I also prefer the current blue scale, but my only real preference is that a slow runtime should not be *red*, as that implies it is somehow "wrong" (especially if a fast runtime is green). Purple-orange-yellow might be fine. I can't mentally visualize how that looks.

samestep commented:
Makes sense. Then in that case @raph5 I think all you need to do is fix Prettier on this PR :)

raph5 commented:
> @raph5 could you clarify what you mean by this? Isn't that exactly how the score is already calculated?

Yes I was meaning the reciprocal of the score then. Though I didn't looked into the computation of the score at the time.

For the colors I was thinking about the one called inferno here but I will try them all to see if there is one I like better.
<img width="586" alt="Screenshot 2025-05-12 at 17 43 15" src="https://github.com/user-attachments/assets/26029197-d8a2-4b9d-8797-788345da20b0" />

Let me change a few things then and I will send the new PR. Can I reorganize a little the website directory?

samestep commented:
Yes go for it!

athas commented:
I don't have the experience to judge the React code, but I really like the way that table looks.

vchuravy commented:
Maybe the only thing I would do is to also plot the "primal" code performance? Since we are comparing across multiple languages with different primal codes, looking at performance of the AD tool requires an anchor against the performance of the primal code.

athas commented:
The primal performance is still plotted in the graphs, but how would you show it in the table?

vchuravy commented:
> The primal performance is still plotted in the graphs, but how would you show it in the table?

:see_no_evil: I had never noticed that you could click on the name of the test and see the curves!

raph5 commented:
I added tensorflow-js to this branch to comply with https://github.com/gradbench/gradbench/pull/561

samestep commented:
@raph5 sorry for being so slow to review! I hope to get to this later today.

As Troels mentioned, though, you'll need to first resolve the merge conflicts, which you can do by merging the `main` branch into your own branch and then pushing.

raph5 commented:
Ok I will merge it later

@athas I replaced that part by ’text-wrap: nowarp’ in the css. Is that ok?

athas commented:
Fine by me.

samestep reviewed (changes requested):
Thanks @raph5 for this PR! Overall I find it to be a minor aesthetic improvement, so, a bit of progress toward #410.

---

For the numbers and symbols that you've added to the grid: I think these can potentially be useful, so I'm interested in them. However, they do have a couple issues that make me a bit unsure:

1. There's now even more coupling between this code and the code for the `repo stats` command in the GradBench CLI, which I've linked previously. I guess this isn't really a big deal since we probably don't expect to change the semantics of these scorers much going forward... but it does add brittleness, so if we _do_ decide to change those later, then it'll be more difficult.

2. There's no text on the website explaining the exact meanings of these symbols or the times. This was kind of already an issue before this PR (@vchuravy has requested text explaining the numerical meanings of the colors), but it becomes accentuated by putting precise numbers on everything.

Of course, it's possible to either ignore or fix both of these issues. But before doing that, I'd want to get a stronger positive signal about whether this is useful to people. From skimming this thread again just now, it seems like @athas is lukewarm on the change, and I don't really see commentary on it from others.

---

Then there are a lot of other changes. It seems like overall this changes the React code to use a somewhat different style than I was using before (e.g. `.then` and `.catch` instead of `async`/`await`), which I'm not necessarily opposed to, but especially judging from the comments in `js/website/src/components/App.tsx`, it's not entirely clear to me how much of this refactoring is an actual improvement. Could you perhaps explain a bit what are the overall improvements you've made, and why specifically they are better than how the code was before?

---

One concrete difference on the website is that, when there's a `?commit=` in the query string, the date picker disappears entirely. I don't find this desirable; I would prefer to keep the old behavior, in which someone viewing a webpage for a specific commit can still use the date picker to instead view nightly data.

A related difference is that the new website seems to always put `?date=` in the query string when the website first loads. I also don't really prefer this; when I wrote the code that's currently in the repo, I intentionally made it so that it only puts `?date=` in the query string when the user selects a different date, and leaves the URL alone if they just stay on the current nightly.

---

I've left a few more specific comments about individual parts of the code as well.

athas commented:
I have warmed to the numbers, and I think they are an improvement. The website lacks descriptions about what things mean, but that was already the case.

raph5 commented:
Thank you for taking the time to review my bulky PR.

When approaching this PR I did my best to keep my changes minimal as you can see by my first 24 lines commit. However, along the way, I accidentally refactored the entire website. The functional changes proposed in this PR are not tied to this refactoring and can absolutely be done in a 50 lines PR.

That said I can defend my choices 😛

---

**First on the functionalities:**

> There's now even more coupling between this code and the code for the `repo stats` command in the GradBench CLI

I agree that that the website is now more coupled with the repo stats command but I think this is desirable. The current grandben.ch implementation is not tightly coupled with the `repo stats` because the colors displayed are not a very precise representation of the data. I think grandben.ch should represent the benchmark data very accurately.
But I think more could be done to make the coupling more explicit. For example taking the reciprocal of the score to get the runtime is quite a sneaky coupling.

> There's no text on the website explaining the exact meanings of these symbols or the times

Agree

---

**Then on the refactoring:**

My focus when refactoring this was to decouple as much as possible the different part of the website. The first thing I did was to centralize the state management at the top of the component hierarchy by moving all the fetching logic in `App.tsx`. This enabled me to represent the UI as independent stateless components.

I removed the `const [state, setState] = useState<State>(...)` as I don't understand the purpose for this abstraction over the useState hook.

I changed the error handling code as I felt the errors were handled two times:
```js
const download = async (prefix: string): Promise<Summary | undefined> => {
  try {
    const response = await fetch(`${prefix}/summary.json`);
    return await response.json();
  } catch (_) {
    return undefined;  // here
  }
};
```
```js
      ) : state.summary.summary === undefined ? (  // and here
```

My changes on the query string were motivated by creating a simple, single source of truth regarding the value of `date` and `commit`. Currently I think the rules regarding the value of these two variables are more complex than they need to be. But I agree that this attempted simplification shouldn't unnecessarily deteriorate the website functionalities.

Finally I rewrote the useEffect in `App.tsx` with the intention of making it less convoluted by focusing on explicitness and exhaustiveness. This is however a subjective measurement.
```js
  // refactored
  const [summary, setSummary] = useState<Summary | null>(null);
  const [summaryStatus, setSummaryStatus] = useState(Status.Loading);
  useEffect(() => {
    setSummaryStatus(Status.Loading);
    downloadSummary(date, commit)
      .then((summary) => {
        setSummary(summary);
        if (summary.date) setDate(summary.date);
        setSummaryStatus(Status.Ok);
      })
      .catch((err) => {
        if (err instanceof NotFoundError) {
          setSummaryStatus(Status.NotFound);
        } else {
          console.error(err);
          setSummaryStatus(Status.Error);
        }
      });
  }, [date, commit, setDate]);
```
```js
  // actual
  const [state, setState] = useState<State>({
    date: parseDate(params.get("date")),
    summary: undefined,
  });
  const prefix = urlPrefix({ commit, date: state.date });
  const downloaded = state.summary?.prefix;
  useEffect(() => {
    // Nothing to do if we've already downloaded this summary.
    if (prefix === downloaded) return;
    (async () => {
      const summary = await download(prefix);
      setState((current) => {
        // Only overwrite the summary if the URL prefix we're using is still the
        // same as the one wanted by the current state.
        if (urlPrefix({ commit, date: current.date }) !== prefix)
          return current;
        const newState: State = { ...current, summary: { prefix, summary } };
        if (current.date === undefined) {
          const date = parseDate(summary?.date);
          // If the user hasn't picked a date and there's no `commit` in the
          // query parameters, then we just downloaded the summary for the most
          // recent nightly build, which probably has a date, so we can just
          // store that to show in the date picker.
          if (date !== undefined) {
            newState.date = date;
          }
        }
        return newState;
      });
    })();
  }, [commit, prefix, downloaded]);
```

I did a lot of other small changes that can be summed up as scattering the code into different files. Those changes are purely subjective and of no real importance to me.

---

If you feel this refactoring is not appropriate I can submit you the functional changes in a minimal line changes.

raph5 commented on js/website/src/components/App.tsx:
```diff
@@ -0,0 +1,120 @@
+  const [summary, setSummary] = useState<Summary | null>(null);
+  const [summaryStatus, setSummaryStatus] = useState(Status.Loading);
+  useEffect(() => {
+    // NOTE: Unnecessary calls to downloadSummary will happen but the already
+    // cumbersome enough without takeing care of that
```
downloadSummary is called each time the useEffect is rerun. On the current implémentation of gradben.ch in production this happens tree times. Though this is not a very big problem as it does not affect loading time
